// generator-angularjs-simple/app/index.js
'use strict';
var util = require('util');
var path = require('path');
var yeoman = require('yeoman-generator');
var yosay = require('yosay');
var chalk = require('chalk');


var AngularjsSimpleGenerator = yeoman.generators.Base.extend({
    init: function () {
        this.pkg = require('../package.json');

        var done = this.async();

        var prompts = [
            {
                name: 'siteName',
                message: 'What is your site\'s name ?',
                default: 'My Angular Site'
            },
            {
                name: 'baseAppName',
                message: 'What is the name of your base angular module?',
                default: 'my_app'
            },
            {
                name: 'authorName',
                message: 'What is your name?',
                default: 'Bill Smith \<billsmith@gmail.com\>'
            },
        ];

        this.prompt(prompts, function (props) {
            this.siteName = props.siteName;
            this.baseAppName = props.baseAppName;
            this.authorName = props.authorName;

            done();
        }.bind(this));

        this.on('end', function () {
            this.installDependencies({
                skipInstall: this.options['skip-install'],
                callback: function() {
                    this.spawnCommand('grunt', ['default'])
                }.bind(this)
            });
        });
    },
    scaffoldFolders: function () {
        this.mkdir('www');
        this.mkdir('www/lib/');
        this.mkdir('www/assets/');
        this.mkdir('www/assets/build/');
        this.mkdir('www/assets/build/css/');
        this.mkdir('www/assets/build/js/');
        this.mkdir('www/assets/img/');
        this.mkdir('www/modules/');
        this.mkdir('www/modules/base/');
        this.mkdir('www/modules/base/coffee/');
        this.mkdir('www/modules/base/sass/');
        this.mkdir('www/modules/base/templates/');
    },
    scaffoldFiles: function() {
        this.template(
            '_index.html',
            'www/index.html',
            {
                'site_name': this.siteName,
                'base_app_name': this.baseAppName
            }
        );

        this.template(
            '_app.coffee',
            'www/modules/base/coffee/app.coffee',
            { 'base_app_name': this.baseAppName }
        );

        this.template(
            '_bower.json',
            'bower.json',
            {
                'app_name': this.baseAppName,
                'app_author': this.authorName
            }
        );

        this.template(
            '_package.json',
            'package.json',
            {
                'app_name': this.baseAppName,
                'app_author': this.authorName
            }
        );

        this.copy('_.bowerrc', '.bowerrc');
        this.copy('_Gruntfile.coffee', 'Gruntfile.coffee');
        this.copy('_config.rb', 'config.rb');
        this.copy('_app.sass', 'www/modules/base/sass/app.sass');
    }
});

module.exports = AngularjsSimpleGenerator;
